contracts/MerkleLib.sol

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+
+
+/**
+ * @notice Library to help with merkle roots, proofs, and claims.
+ */
+library MerkleLib {
+    // TODO: some of these data structures can be moved out if convenient.
+    // This data structure is used in the settlement process on L1.
+    // Each PoolRebalance structure is responsible for balancing a single chain's SpokePool across all tokens.
+    struct PoolRebalance {
+        // Used as the index in the bitmap to track whether this leaf has been executed or not.
+        uint256 leafId;
+        // This is used to know which chain to send cross-chain transactions to (and which SpokePool to sent to).
+        uint256 chainId;
+        // This only needs to be emitted in an event so future proposals know the most recent block number.
+        uint256 endBlock;
+
+        // The following arrays are required to be the same length.
+        address[] tokenAddresses; // All whitelisted tokens for this chain in the order of whitelisting.
+        uint256[] bundleLpFees; // LP fee amount per token in this bundle.
+
+        // This array is grouped with the two above, and it represents the amount to send or request back from the
+        // SpokePool. If positive, the pool needs to pay the SpokePool. If negative the SpokePool needs to pay the
+        // HubPool. There can be arbitrarily complex rebalancing rules defined offchain. This number is only nonzero
+        // when the rules indicate that a rebalancing action should occur. This means that the DestinationDistribution
+        // associated with this PoolRebalance can indicate a balance change, but that the HubPool does not rebalance.
+        // When a rebalance does occur, the netSendAmount is the sum of all netBalanceChange values in the
+        // DestinationDistribution structs since the last rebalance action.
+        int256[] netSendAmount;
+    }
+
+    // This leaf is meant to be decoded in the SpokePool in order to pay out individual relayers for this bundle.
+    struct DestinationDistribution {
+        // Used as the index in the bitmap to track whether this leaf has been executed or not.
+        uint256 leafId;
+        // Used to verify that this is being decoded on the correct chainId.
+        uint256 chainId;
+        // This is the amount to return to the HubPool. This occurs when there is a PoolRebalance netSendAmount that is
+        // negative. This is just that value inverted. 
```

Comment:
I'm not sure if we want the inverted amount. it might be easier if this is always a int256. a positive number always means L1->L2 flow and a negative number always means L2->L1 flow. dont bake any polarity changes into the variable.

Reply:
As discussed today, we would only set this number if we wanted to tell the receiving chain to send tokens back to mainnet. So I made this positive since there is no need to communicate when tokens are being sent _to_ the L2 (aka the opposite sign).